llvm/test/CodeGen/RISCV/cheri/compress.ll

```diff
@@ -0,0 +1,106 @@
+; RUN: cat %s > %t.fnattr
+; RUN: echo 'attributes #0 = { nounwind "target-features"="+c,+xcheri,+cap-mode,+xcheri-norvc" }' >> %t.fnattr
+; RUN: %riscv64_cheri_purecap_llc -filetype=obj < %t.fnattr \
+; RUN:   | llvm-objdump -d --mattr=+c -M no-aliases - | FileCheck %s --check-prefix=CHECK-NORVC
+
+; Basic check that we can use CHERI compressed instructions
+
+define i32 @loadstore(i32 addrspace(200)* %intptrarg, i8 addrspace(200)* addrspace(200)* %ptrptrarg) addrspace(200) #0 {
+; CHECK-LABEL: <loadstore>:
+; CHECK-NEXT:    c.cincoffset16csp csp, -32
+; CHECK-NEXT:    c.clw a2, 0(ca0)
+; CHECK-NEXT:    c.li a3, 1
+; CHECK-NEXT:    c.csw a3, 0(ca0)
+; CHECK-NEXT:    c.clc ca0, 0(ca1)
+; CHECK-NEXT:    c.csc ca0, 0(ca1)
+; CHECK-NEXT:    c.csccsp ca0, 16(csp)
+; CHECK-NEXT:    c.clccsp ca0, 16(csp)
+; CHECK-NEXT:    c.cswcsp a2, 0(csp)
+; CHECK-NEXT:    c.clwcsp a0, 0(csp)
+; CHECK-NEXT:    c.mv a0, a2
+; CHECK-NEXT:    c.cincoffset16csp csp, 32
+; CHECK-NEXT:    c.cjr cra
+; CHECK-NORVC-LABEL: <loadstore>:
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}cincoffset csp, csp, -32
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}clw a2, 0(ca0)
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}c.li a3, 1
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}csw a3, 0(ca0)
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}clc ca0, 0(ca1)
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}csc ca0, 0(ca1)
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}csc ca0, 16(csp)
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}clc ca0, 16(csp)
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}csw a2, 0(csp)
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}clw a0, 0(csp)
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}c.mv a0, a2
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}cincoffset csp, csp, 32
+; CHECK-NORVC-NEXT:  {{[^a-z.]}}cjalr cnull, 0(cra)
+  %stackptr = alloca i8 addrspace(200)*, align 16, addrspace(200)
+  %stackint = alloca i32, align 16, addrspace(200)
+  %val = load volatile i32, i32 addrspace(200)* %intptrarg
+  store volatile i32 1, i32 addrspace(200)* %intptrarg
+  %ptrval = load volatile i8 addrspace(200)*, i8 addrspace(200)* addrspace(200)* %ptrptrarg
+  store volatile i8 addrspace(200)* %ptrval, i8 addrspace(200)* addrspace(200)* %ptrptrarg
+  store volatile i8 addrspace(200)* %ptrval, i8 addrspace(200)* addrspace(200)* %stackptr
+  %stackptrval = load volatile i8 addrspace(200)*, i8 addrspace(200)* addrspace(200)* %stackptr
+  store volatile i32 %val, i32 addrspace(200)* %stackint
+  %stackintval = load volatile i32, i32 addrspace(200)* %stackint
+  ret i32 %val
+}
+
+define i32 @call() addrspace(200) #0 {
+; CHECK-LABEL: <call>:
+; CHECK-NEXT:    c.cincoffset16csp csp, -{{112|144}}
+; CHECK-NEXT:    c.csccsp cra, {{104|128}}(csp)
+; CHECK-NEXT:    c.cincoffset4cspn ca0, csp, 64
+; CHECK-NEXT:    csetbounds ca0, ca0, {{32|64}}
+; CHECK-NEXT:    cincoffset ca1, ca0, {{24|48}}
+; CHECK-NEXT:    cincoffset ca0, csp, 0
+; CHECK-NEXT:    csetbounds ca0, ca0, 64
+; CHECK-NEXT:    cincoffset ca0, ca0, 12
+; NB: No c.cjalr here, linker relaxations expect the full sequence
```

Comment:
Ugh, comments in the middle of CHECK lines are really hard to see (and comments that aren't CHECK lines should be `;;` anyway), please hoist this to above the IR definition